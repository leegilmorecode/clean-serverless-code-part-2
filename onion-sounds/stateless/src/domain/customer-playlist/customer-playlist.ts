import * as customerPlaylistCreatedEvent from '@events/customer-playlist-created';
import * as songAddedToPlaylistEvent from '@events/song-added-to-playlist';

import {
  CreateCustomerPlaylistProps,
  NewCustomerPlaylistProps,
} from '@models/customer-playlist-types';

import { CustomerPlaylistDto } from '@dto/customer-playlist';
import { Entity } from '@entity/entity';
import { MaxPlaylistSizeError } from '@errors/max-playlist-size-error';
import { schema } from '@schemas/customer-playlist.schema';

/**
 * Customer Playlist Domain Entity
 *
 * A customer playlist is an entity which allows customers to manage one or more playlists.
 */
export class CustomerPlaylist extends Entity<CreateCustomerPlaylistProps> {
  private constructor({
    id,
    created,
    updated,
    ...props
  }: CreateCustomerPlaylistProps) {
    super(props, id, created, updated);
  }

  // create an instance of a new customer playlist through a factory method
  public static createPlaylist(
    props: NewCustomerPlaylistProps
  ): CustomerPlaylist {
    const customerAccountProps: CreateCustomerPlaylistProps = {
      playlistName: props.playlistName,
      songIds: [],
    };

    const instance: CustomerPlaylist = new CustomerPlaylist(
      customerAccountProps
    );

    instance.validate(schema);

    instance.addDomainEvent({
      event: instance.toDto(),
      eventName: customerPlaylistCreatedEvent.eventName,
      source: customerPlaylistCreatedEvent.eventSource,
      eventSchema: customerPlaylistCreatedEvent.eventSchema,
      eventVersion: customerPlaylistCreatedEvent.eventVersion,
    });

    return instance;
  }

  // add a song to the playlist
  public addSongToPlaylist(songId: string): void {
    if (this.props.songIds.length >= 4) {
      throw new MaxPlaylistSizeError('the maximum playlist length is 4');
    }

    this.props.songIds.push(songId);

    this.setUpdatedDate();
    this.validate(schema);

    this.addDomainEvent({
      event: this.toDto(),
      eventName: songAddedToPlaylistEvent.eventName,
      source: songAddedToPlaylistEvent.eventSource,
      eventSchema: songAddedToPlaylistEvent.eventSchema,
      eventVersion: songAddedToPlaylistEvent.eventVersion,
    });
  }

  // create a dto based on the domain instance
  public toDto(): CustomerPlaylistDto {
    return {
      id: this.id,
      created: this.created,
      updated: this.updated,
      playlistName: this.props.playlistName,
      songIds: this.props.songIds,
    };
  }

  // create a domain object based on the dto
  public static toDomain(raw: CustomerPlaylistDto): CustomerPlaylist {
    const instance = new CustomerPlaylist(raw);
    return instance;
  }
}
