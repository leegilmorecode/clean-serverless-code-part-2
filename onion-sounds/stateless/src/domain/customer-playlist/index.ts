export { CustomerPlaylist } from './customer-playlist';
