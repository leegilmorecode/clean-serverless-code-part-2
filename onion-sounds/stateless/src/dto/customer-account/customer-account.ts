import {
  PaymentStatus,
  SubscriptionType,
} from '@models/customer-account-types';

import { CustomerAddress } from '@domain/customer-address/customer-address';
import { CustomerAddressProps } from '@models/customer-address-types';
import { CustomerPlaylistProps } from '@models/customer-playlist-types';

export type CreateCustomerAccountDto = {
  id?: string;
  created?: string;
  updated?: string;
  firstName: string;
  surname: string;
  subscriptionType: SubscriptionType;
  paymentStatus: PaymentStatus;
  customerAddress: CustomerAddress;
};

export type CustomerAccountDto = {
  id: string;
  created: string;
  updated: string;
  firstName: string;
  surname: string;
  subscriptionType: SubscriptionType;
  paymentStatus: PaymentStatus;
  playlists: CustomerPlaylistProps[];
  customerAddress: CustomerAddressProps;
};

export type NewCustomerAccountDto = {
  firstName: string;
  surname: string;
  customerAddress: CustomerAddressProps;
};
