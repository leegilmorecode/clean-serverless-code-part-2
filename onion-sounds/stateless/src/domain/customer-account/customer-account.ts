import * as customerAccountCreatedEvent from '@events/customer-account-created';
import * as customerAccountUpdatedEvent from '@events/customer-account-updated';
import * as customerAccountUpgradedEvent from '@events/customer-account-upgraded';

import {
  CreateCustomerAccountProps,
  NewCustomerAccountProps,
  PaymentStatus,
  SubscriptionType,
} from '@models/customer-account-types';
import {
  CustomerPlaylistProps,
  NewCustomerPlaylistProps,
} from '@models/customer-playlist-types';

import { AggregateRoot } from '@entity/aggregate-root';
import { CustomerAccountDto } from '@dto/customer-account';
import { CustomerAddress } from '@domain/customer-address/customer-address';
import { CustomerPlaylist } from '@domain/customer-playlist';
import { DomainEvent } from '@entity/domain-event';
import { MaxNumberOfPlaylistsError } from '@errors/max-number-of-playlists-error';
import { PaymentInvalidError } from '@errors/payment-invalid-error';
import { PlaylistNotFoundError } from '@errors/playlist-not-found-error';
import { SubscriptionAlreadyUpgradedError } from '@errors/subscription-already-upgraded-error';
import { schema } from '@schemas/customer-account.schema';

// only works with entities and value objects - calling the repositories
// adapter --> use case --> (domain) <-- repository <-- adapter

/**
 * Customer Account Domain Entity
 *
 * A customer account is an entity which allows customers to create a new account, and upgrade their
 * account subscription.
 */
export class CustomerAccount extends AggregateRoot<CreateCustomerAccountProps> {
  private constructor({
    id,
    created,
    updated,
    ...props
  }: CreateCustomerAccountProps) {
    super(props, id, created, updated);
  }

  // retrieve all domain events in the aggregate inc children
  // adapter --> use case --> (domain)
  public retrieveDomainEvents(): DomainEvent[] {
    const childEvents: DomainEvent[] = this.props.playlists
      .map((playList: CustomerPlaylist) => playList.domainEvents)
      .reduce(
        (domainEvent: DomainEvent[], item: DomainEvent[]) =>
          domainEvent.concat(item),
        []
      );

    const existingEvents = [...this.domainEvents.concat(childEvents, [])];

    // clear down the existing events on the queue
    this.clearDomainEvents();
    this.props.playlists.forEach((playlist) => playlist.clearDomainEvents());

    return existingEvents;
  }

  // retrieve an instance of the playlist on a given account
  // adapter --> use case --> (domain)
  public retrievePlaylist(playlistId: string): CustomerPlaylist {
    const playlist: CustomerPlaylist | undefined = this.props.playlists.find(
      (playlist: CustomerPlaylist) => playlist.id === playlistId
    );

    if (playlist === undefined) {
      throw new PlaylistNotFoundError(`Playlist ${playlistId} not found`);
    }
    return playlist;
  }

  // create an instance of a new customer account through a factory method
  // adapter --> use case --> (domain)
  public static createAccount(props: NewCustomerAccountProps): CustomerAccount {
    const customerAccountProps: CreateCustomerAccountProps = {
      firstName: props.firstName,
      surname: props.surname,
      subscriptionType: SubscriptionType.Basic,
      paymentStatus: PaymentStatus.Valid,
      playlists: [],
      customerAddress: CustomerAddress.create(props.customerAddress),
    };

    const instance: CustomerAccount = new CustomerAccount(customerAccountProps);
    instance.validate(schema);

    instance.addDomainEvent({
      event: instance.toDto(),
      eventName: customerAccountCreatedEvent.eventName,
      source: customerAccountCreatedEvent.eventSource,
      eventSchema: customerAccountCreatedEvent.eventSchema,
      eventVersion: customerAccountCreatedEvent.eventVersion,
    });

    return instance;
  }

  // upgrade a customer account from basic to advanced through a service (use case)
  // adapter --> use case --> (domain)
  public upgradeAccount(): void {
    // only allow an upgrade if the payment status is valid
    if (this.props.paymentStatus === PaymentStatus.Invalid) {
      throw new PaymentInvalidError('Payment is invalid - unable to upgrade');
    }

    // we can not upgrade an account which is already upgraded
    if (this.props.subscriptionType === SubscriptionType.Upgraded) {
      throw new SubscriptionAlreadyUpgradedError(
        'Subscription is already upgraded - unable to upgrade'
      );
    }

    // update the account to upgraded
    this.props.subscriptionType = SubscriptionType.Upgraded;

    this.setUpdatedDate();
    this.validate(schema);

    this.addDomainEvent({
      event: this.toDto(),
      eventName: customerAccountUpgradedEvent.eventName,
      source: customerAccountUpgradedEvent.eventSource,
      eventSchema: customerAccountUpgradedEvent.eventSchema,
      eventVersion: customerAccountUpgradedEvent.eventVersion,
    });
  }

  // create a customer playlist on the customer account through a service (use case)
  // adapter --> use case --> (domain)
  public createPlaylist(playlistName: string): CustomerPlaylist {
    const playlistProps: NewCustomerPlaylistProps = { playlistName };
    const newPlaylist: CustomerPlaylist =
      CustomerPlaylist.createPlaylist(playlistProps);

    if (this.props.playlists.length === 2) {
      throw new MaxNumberOfPlaylistsError(
        'maximum number of playlists reached'
      );
    }

    // create the new playlist
    this.props.playlists.push(newPlaylist);

    this.setUpdatedDate();
    this.validate(schema);

    this.addDomainEvent({
      event: this.toDto(),
      eventName: customerAccountUpdatedEvent.eventName,
      source: customerAccountUpdatedEvent.eventSource,
      eventSchema: customerAccountUpdatedEvent.eventSchema,
      eventVersion: customerAccountUpdatedEvent.eventVersion,
    });

    return newPlaylist;
  }

  // create a dto based on the domain instance
  public toDto(): CustomerAccountDto {
    return {
      id: this.id,
      created: this.created,
      updated: this.updated,
      firstName: this.props.firstName,
      surname: this.props.surname,
      subscriptionType: this.props.subscriptionType,
      paymentStatus: this.props.paymentStatus,
      playlists: this.props.playlists.map((item: CustomerPlaylist) =>
        item.toDto()
      ),
      customerAddress: this.props.customerAddress.toDto(),
    };
  }

  // create a domain object based on the dto
  public static toDomain(raw: CustomerAccountDto): CustomerAccount {
    const playlists: CustomerPlaylist[] = raw.playlists.map(
      (item: CustomerPlaylistProps) => CustomerPlaylist.toDomain(item)
    );

    const customerAddress = CustomerAddress.create(raw.customerAddress);

    const instance = new CustomerAccount({
      ...raw,
      playlists,
      customerAddress,
    });
    return instance;
  }
}
