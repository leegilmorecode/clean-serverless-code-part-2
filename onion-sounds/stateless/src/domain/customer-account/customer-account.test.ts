import {
  NewCustomerAccountProps,
  PaymentStatus,
  SubscriptionType,
} from '@models/customer-account-types';

import { CustomerAccount } from '@domain/customer-account';
import { CustomerAccountDto } from '@dto/customer-account';
import { CustomerPlaylist } from '@domain/customer-playlist';

let customer: CustomerAccountDto = {
  created: '2022-01-01T00:00:00.000Z',
  firstName: 'Lee',
  id: 'f39e49ad-8f88-448f-8a15-41d560ad6d70',
  paymentStatus: PaymentStatus.Valid,
  subscriptionType: SubscriptionType.Basic,
  surname: 'Gilmore',
  updated: '2022-01-01T00:00:00.000Z',
  playlists: [],
  customerAddress: {
    addressLineOne: 'line one',
    addressLineTwo: 'line two',
    addressLineThree: 'line three',
    addressLineFour: 'line four',
    addressLineFive: 'line five',
    postCode: 'ne11bb',
  },
};

describe('customer-account', () => {
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date('2022-01-01'));
  });

  afterAll(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
  });

  describe('create account', () => {
    it('should fail creating a new account with an invalid customer', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Â±', // invalid
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act / assert
      expect(() =>
CustomerAccount.createAccount(newCustomerAccount)).
toThrowErrorMatchingInlineSnapshot(`"[{\\"instancePath\\":\\"/firstName\\",\\"schemaPath\\":\\"#/properties/firstName/pattern\\",\\"keyword\\":\\"pattern\\",\\"params\\":{\\"pattern\\":\\"^[a-zA-Z]+$\\"},\\"message\\":\\"must match pattern \\\\\\"^[a-zA-Z]+$\\\\\\"\\"}]"`);
    });

    it('should create the new account successfully with a valid customer', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer).toMatchSnapshot();
    });
  });

  describe('upgrade account', () => {
    it('should upgrade an existing customer by id if payment status is valid', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        subscriptionType: SubscriptionType.Basic,
      });

      // act
      customerAccount.upgradeAccount();

      // assert
      expect(customerAccount).toMatchSnapshot();
    });

    it('should not upgrade an existing customer by id if payment status is invalid', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        paymentStatus: PaymentStatus.Invalid,
      });

      // act / assert
      expect(() => customerAccount.upgradeAccount()).toThrow(
        'Payment is invalid - unable to upgrade'
      );
    });

    it('should not upgrade an existing customer by id if status is already upgraded', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        paymentStatus: PaymentStatus.Valid,
        subscriptionType: SubscriptionType.Upgraded,
      });

      // act / assert
      expect(() => customerAccount.upgradeAccount()).toThrow(
        'Subscription is already upgraded - unable to upgrade'
      );
    });
  });

  describe('create playlist', () => {
    it('should add the playlist successfully', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        subscriptionType: SubscriptionType.Basic,
      });
      const playlistName = 'newplaylist';

      // act
      customerAccount.createPlaylist(playlistName);

      // assert
      expect(customerAccount).toMatchSnapshot();
    });

    it('should throw an error if there are more than 2 playlists', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        subscriptionType: SubscriptionType.Basic,
      });

      customerAccount.createPlaylist('playlistOne');
      customerAccount.createPlaylist('playlistTwo');

      // act & assert
      expect(() =>
        customerAccount.createPlaylist('playlistThree')
      ).toThrowErrorMatchingInlineSnapshot(
        `"maximum number of playlists reached"`
      );
    });
  });

  describe('toDto', () => {
    it('should create the correct dto', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer.toDto()).toMatchSnapshot();
    });
  });

  describe('toDomain', () => {
    it('should create a domain object based on a dto', () => {
      // arrange
      const customerAccountProps: CustomerAccountDto = {
        id: 'f39e49ad-8f88-448f-8a15-41d560ad6d70',
        firstName: 'Lee',
        surname: 'Gilmore',
        paymentStatus: PaymentStatus.Valid,
        subscriptionType: SubscriptionType.Upgraded,
        created: '2022-01-01T00:00:00.000Z',
        updated: '2022-01-01T00:00:00.000Z',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
        playlists: [
          {
            id: 'a49e49ad-8f88-448f-8a15-41d560ad6d70',
            playlistName: 'testplaylist',
            songIds: [],
            created: '2022-01-01T00:00:00.000Z',
            updated: '2022-01-01T00:00:00.000Z',
          },
        ],
      };

      // act
      const customerAccount: CustomerAccount =
        CustomerAccount.toDomain(customerAccountProps);

      // assert
      expect(customerAccount).toMatchSnapshot();
    });
  });

  describe('retrieve playlist', () => {
    it('should return the playlist successfully if found', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);
      const createdPlaylist: CustomerPlaylist =
        newCustomer.createPlaylist('playlistTwo');

      // assert
      expect(newCustomer.retrievePlaylist(createdPlaylist.id)).
toMatchInlineSnapshot(`
CustomerPlaylist {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "playlistName": "playlistTwo",
        "songIds": Array [],
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerPlaylistCreated",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "playlistName": "playlistTwo",
    "songIds": Array [],
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`);
    });

    it('should throw an error if playlist not found', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);
      newCustomer.createPlaylist('playlistone');

      // assert
      expect(() =>
        newCustomer.retrievePlaylist('playlisttwo')
      ).toThrowErrorMatchingInlineSnapshot(`"Playlist playlisttwo not found"`);
    });
  });

  describe('add domain event', () => {
    it('should return the domain events on the object', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);
      newCustomer.createPlaylist('playlistone');

      // assert
      expect(newCustomer.retrieveDomainEvents()).toMatchSnapshot();
    });

    it('should clear all domain events once retrieved', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
        customerAddress: {
          addressLineOne: 'line one',
          addressLineTwo: 'line two',
          addressLineThree: 'line three',
          addressLineFour: 'line four',
          addressLineFive: 'line five',
          postCode: 'ne11bb',
        },
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);
      newCustomer.createPlaylist('playlistone');
      newCustomer.retrieveDomainEvents(); // call this once to flush the events

      // assert
      expect(newCustomer.retrieveDomainEvents()).toEqual([]);
    });
  });
});
