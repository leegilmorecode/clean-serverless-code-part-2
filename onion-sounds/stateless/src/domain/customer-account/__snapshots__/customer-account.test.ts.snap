// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customer-account add domain event should return the domain events on the object 1`] = `
Array [
  Object {
    "event": Object {
      "created": "2022-01-01T00:00:00.000Z",
      "customerAddress": Object {
        "addressLineFive": "line five",
        "addressLineFour": "line four",
        "addressLineOne": "line one",
        "addressLineThree": "line three",
        "addressLineTwo": "line two",
        "postCode": "ne11bb",
      },
      "firstName": "Lee",
      "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
      "paymentStatus": "Valid",
      "playlists": Array [],
      "subscriptionType": "Basic",
      "surname": "Gilmore",
      "updated": "2022-01-01T00:00:00.000Z",
    },
    "eventDateTime": "2022-01-01T00:00:00.000Z",
    "eventName": "CustomerAccountCreated",
    "eventVersion": "1",
    "source": "com.customer-account-onion",
  },
  Object {
    "event": Object {
      "created": "2022-01-01T00:00:00.000Z",
      "customerAddress": Object {
        "addressLineFive": "line five",
        "addressLineFour": "line four",
        "addressLineOne": "line one",
        "addressLineThree": "line three",
        "addressLineTwo": "line two",
        "postCode": "ne11bb",
      },
      "firstName": "Lee",
      "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
      "paymentStatus": "Valid",
      "playlists": Array [
        Object {
          "created": "2022-01-01T00:00:00.000Z",
          "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
          "playlistName": "playlistone",
          "songIds": Array [],
          "updated": "2022-01-01T00:00:00.000Z",
        },
      ],
      "subscriptionType": "Basic",
      "surname": "Gilmore",
      "updated": "2022-01-01T00:00:00.000Z",
    },
    "eventDateTime": "2022-01-01T00:00:00.000Z",
    "eventName": "CustomerAccountUpdated",
    "eventVersion": "1",
    "source": "com.customer-account-onion",
  },
  Object {
    "event": Object {
      "created": "2022-01-01T00:00:00.000Z",
      "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
      "playlistName": "playlistone",
      "songIds": Array [],
      "updated": "2022-01-01T00:00:00.000Z",
    },
    "eventDateTime": "2022-01-01T00:00:00.000Z",
    "eventName": "CustomerPlaylistCreated",
    "eventVersion": "1",
    "source": "com.customer-account-onion",
  },
]
`;

exports[`customer-account create account should create the new account successfully with a valid customer 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "customerAddress": Object {
          "addressLineFive": "line five",
          "addressLineFour": "line four",
          "addressLineOne": "line one",
          "addressLineThree": "line three",
          "addressLineTwo": "line two",
          "postCode": "ne11bb",
        },
        "firstName": "Lee",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "paymentStatus": "Valid",
        "playlists": Array [],
        "subscriptionType": "Basic",
        "surname": "Gilmore",
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerAccountCreated",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "customerAddress": CustomerAddress {
      "props": Object {
        "addressLineFive": "line five",
        "addressLineFour": "line four",
        "addressLineOne": "line one",
        "addressLineThree": "line three",
        "addressLineTwo": "line two",
        "postCode": "ne11bb",
      },
    },
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "playlists": Array [],
    "subscriptionType": "Basic",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account create playlist should add the playlist successfully 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "customerAddress": Object {
          "addressLineFive": "line five",
          "addressLineFour": "line four",
          "addressLineOne": "line one",
          "addressLineThree": "line three",
          "addressLineTwo": "line two",
          "postCode": "ne11bb",
        },
        "firstName": "Lee",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "paymentStatus": "Valid",
        "playlists": Array [
          Object {
            "created": "2022-01-01T00:00:00.000Z",
            "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
            "playlistName": "newplaylist",
            "songIds": Array [],
            "updated": "2022-01-01T00:00:00.000Z",
          },
        ],
        "subscriptionType": "Basic",
        "surname": "Gilmore",
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerAccountUpdated",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "customerAddress": CustomerAddress {
      "props": Object {
        "addressLineFive": "line five",
        "addressLineFour": "line four",
        "addressLineOne": "line one",
        "addressLineThree": "line three",
        "addressLineTwo": "line two",
        "postCode": "ne11bb",
      },
    },
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "playlists": Array [
      CustomerPlaylist {
        "_created": "2022-01-01T00:00:00.000Z",
        "_domainEvents": Array [
          Object {
            "event": Object {
              "created": "2022-01-01T00:00:00.000Z",
              "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
              "playlistName": "newplaylist",
              "songIds": Array [],
              "updated": "2022-01-01T00:00:00.000Z",
            },
            "eventDateTime": "2022-01-01T00:00:00.000Z",
            "eventName": "CustomerPlaylistCreated",
            "eventVersion": "1",
            "source": "com.customer-account-onion",
          },
        ],
        "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "_updated": "2022-01-01T00:00:00.000Z",
        "props": Object {
          "created": "2022-01-01T00:00:00.000Z",
          "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
          "playlistName": "newplaylist",
          "songIds": Array [],
          "updated": "2022-01-01T00:00:00.000Z",
        },
      },
    ],
    "subscriptionType": "Basic",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account toDomain should create a domain object based on a dto 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "customerAddress": CustomerAddress {
      "props": Object {
        "addressLineFive": "line five",
        "addressLineFour": "line four",
        "addressLineOne": "line one",
        "addressLineThree": "line three",
        "addressLineTwo": "line two",
        "postCode": "ne11bb",
      },
    },
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "playlists": Array [
      CustomerPlaylist {
        "_created": "2022-01-01T00:00:00.000Z",
        "_domainEvents": Array [],
        "_id": "a49e49ad-8f88-448f-8a15-41d560ad6d70",
        "_updated": "2022-01-01T00:00:00.000Z",
        "props": Object {
          "created": "2022-01-01T00:00:00.000Z",
          "id": "a49e49ad-8f88-448f-8a15-41d560ad6d70",
          "playlistName": "testplaylist",
          "songIds": Array [],
          "updated": "2022-01-01T00:00:00.000Z",
        },
      },
    ],
    "subscriptionType": "Upgraded",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account toDto should create the correct dto 1`] = `
Object {
  "created": "2022-01-01T00:00:00.000Z",
  "customerAddress": Object {
    "addressLineFive": "line five",
    "addressLineFour": "line four",
    "addressLineOne": "line one",
    "addressLineThree": "line three",
    "addressLineTwo": "line two",
    "postCode": "ne11bb",
  },
  "firstName": "Lee",
  "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "paymentStatus": "Valid",
  "playlists": Array [],
  "subscriptionType": "Basic",
  "surname": "Gilmore",
  "updated": "2022-01-01T00:00:00.000Z",
}
`;

exports[`customer-account upgrade account should upgrade an existing customer by id if payment status is valid 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "customerAddress": Object {
          "addressLineFive": "line five",
          "addressLineFour": "line four",
          "addressLineOne": "line one",
          "addressLineThree": "line three",
          "addressLineTwo": "line two",
          "postCode": "ne11bb",
        },
        "firstName": "Lee",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "paymentStatus": "Valid",
        "playlists": Array [],
        "subscriptionType": "Upgraded",
        "surname": "Gilmore",
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerAccountUpgraded",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "customerAddress": CustomerAddress {
      "props": Object {
        "addressLineFive": "line five",
        "addressLineFour": "line four",
        "addressLineOne": "line one",
        "addressLineThree": "line three",
        "addressLineTwo": "line two",
        "postCode": "ne11bb",
      },
    },
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "playlists": Array [],
    "subscriptionType": "Upgraded",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;
